[TAN-4829] Add basic `File` model

adessy opened the pull request:
# Changelog
## Added
- [TAN-4829] Implement the new `File` model and its CRUD API.

adessy committed:
[TAN-4829] Add basic `File` model

This model will eventually replace and unify most of the other file models, such as `PhaseFile`, `EventFile`, `StaticPageFile`, and others.

adessy committed:
[TAN-4829] Add basic `GET .../files` API endpoint

Implemented a new `.../files` endpoint. The (new unified version of) files is currently accessible only to admins.

adessy committed:
[TAN-4829] Add basic `GET .../files/:id` API endpoint

adessy committed:
[TAN-4829] Add basic `POST .../files` API endpoint

adessy committed:
[TAN-4829] Add basic `DELETE .../files` API endpoint

adessy committed:
[TAN-4829] Prevent admins from uploading files on behalf of other users

adessy committed:
[TAN-4829] Nullify file uploader when the user is deleted

adessy committed:
[TAN-4829 Add side effects to file creation and deletion

adessy committed:
[TAN-4829] Store the file `size` in the `File` model

We now cache the file size directly in the model to allow the back end to sort and filter files by size without triggering requests to the storage back end (S3).

For now, the field isn’t required, as we believe this may simplify the migration of the other file models into the new `Files::File`.

adessy committed:
[TAN-4829] Use smaller pdf fixture for the `files` factory

adessy committed:
[TAN-4829] Store the file `mime_type` in the `File` model

adessy committed:
[TAN-4829] Make the `files` factory more consistent and convenient

Now dynamically resolves the file fixture based on the file name.

adessy committed:
[TAN-4829] Add uploader filter to `GET .../files` API endpoint

adessy committed:
[TAN-4829] Add sorting to `GET .../files` API endpoint

Allows files to be sorted by the following attributes: `created_at`, `name`, and `size`, or any combination of these.

adessy committed:
[TAN-4829] Add project associations and filtering to `Files::File`

Each project has its own file space, even though files can be shared across multiple projects. Technically, a file can exist without being associated with any project, but that won't be used for now.

adessy committed:
[TAN-4829] Support multiple values for `uploader` and `project` filters in `GET .../files` API endpoint

jamesspeake commented on back/app/controllers/web_api/v1/files_v2_controller.rb:
```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+class WebApi::V1::FilesV2Controller < ApplicationController
```
I'm not sure about the use of `FilesV2` as a name (particularly when you see it in the context of `web_api/v1`, but I cannot think of anything more appropriate

jamesspeake replied:
Just thought, could this be name spaced in the same way as the model? `WebApi::v1::Files::FilesController`?

adessy replied:
Good point! I didn’t think too much about it because my initial goal was for it to replace the existing `FilesController` by the end of the tandem cycle.

jamesspeake commented on back/app/controllers/web_api/v1/files_v2_controller.rb:
```diff
@@ -0,0 +1,91 @@
+    raise "Unsupported sort attribute(s): #{unsupported_attrs.join(', ')}" if unsupported_attrs.present?
+
+    order_params
+  end
```
Do we do this style of sorting elsewhere? If so, it feels like much of this method should/could be reusable

adessy replied:
Yes, I mentioned it briefly during the bet showcase. I'll maybe try to make it a service in another PR, I didn’t want to clutter this one. Also, as it stands, it only works with the direct attributes of the model (not transformations or attributes of associations).

adessy replied:
Logged: https://www.notion.so/govocal/Introduce-an-OrderService-for-multi-sort-2239663b7b268036a7ebec0e3888bcf7

jamesspeake commented on back/app/models/files/file.rb:
```diff
@@ -0,0 +1,51 @@
+#
+#  fk_rails_...  (uploader_id => users.id)
+#
+module Files
```
Wonder if a comment here would be useful to help differentiate this namespaces `File` from existing files

adessy replied:
Logged: https://www.notion.so/govocal/Document-Files-module-2239663b7b2680c1942ce270bef820cd

jamesspeake commented on back/app/models/files/file.rb:
```diff
@@ -0,0 +1,51 @@
+    has_many :files_projects, class_name: 'Files::FilesProject', dependent: :destroy
+    has_many :projects, through: :files_projects
+
+    # TODO: Maybe reconsider the name of this column.
+    # TODO: Using temporarily the ProjectFolders::FileUploader
```
Does this need sorting?

adessy replied:
Already did it, at least partially: https://github.com/CitizenLabDotCo/citizenlab/pull/11450

jamesspeake commented on back/app/policies/files/file_policy.rb:
```diff
@@ -0,0 +1,43 @@
+
+      # Allow creation only if the user moderates all associated projects.
+      #
+      # Note: w (instead of +record.projects+)
```
What does `w` refer to here? Not sure I understand this comment

adessy replied:
`w` is probably me using a shortcut with the focus on the wrong window ^^

adessy replied:
Fixed: 
```ruby
# Note: Getting the projects by querying `Project` directly instead of using
# `record.projects` because +record.files_projects+ may not be persisted yet
# (since this is a `create` action), which can cause issues with some SQL queries.
```

jamesspeake reviewed (approved):
Looks good to me. My only slight struggle is the naming and how it is the same but different to other files. But I'm not sure I have any suggestion for that.

adessy committed:
[TAN-4829] Rename and restructure FilesV2Controller to Files::FilesController

Comment from the PR review: https://github.com/CitizenLabDotCo/citizenlab/pull/11388

Co-authored-by: James Speake <jamesspeake@users.noreply.github.com>
Co-authored-by: Adrien Dessy <209268+adessy@users.noreply.github.com>

adessy committed:
[TAN-4829] Fix comment in `file_policy.rb`

adessy committed:
Merge branch 'master' into TAN-4829/files-endpoints

# Conflicts:
#	back/db/structure.sql
